### PR TITLE
feat(ai-rate-limiting): add expression-based limit strategy

### PR DESCRIPTION
**Pull request by nic-6443**
### Description

This PR adds the `expression` limit strategy to the `ai-rate-limiting` plugin.

#### Expression strategy

The `expression` limit strategy allows defining rate limit groups using [lua-resty-expr](https://github.com/api7/lua-resty-expr) expressions. Each group can have its own `count`, `time_window`, and matching `expression`. When a request matches multiple groups, the **first matching group** is used. If no group matches, the request is passed through without rate limiting.

This enables fine-grained AI token rate limiting based on request attributes (headers, query params, variables, etc.).

#### Example config

```json
{
  "limit_strategy": "expression",
  "cost_expr": "input_tokens + completion_tokens",
  "limit_groups": [
    {
      "expression": [["http_x_model", "==", "gpt-4"]],
      "count": 500,
      "time_window": 60
    },
    {
      "expression": [["http_x_model", "==", "gpt-3.5"]],
      "count": 1000,
      "time_window": 60
    }
  ]
}
```

### Checklist

- [x] Expression strategy implementation with schema, compile, eval functions
- [x] Handle edge cases: negative, NaN, Inf token usage
- [x] Safe expression environment (upstream usage keys cannot shadow built-in functions)
- [x] 13 test cases covering expression rate limiting scenarios

### Note on remaining header accuracy

The `X-AI-RateLimit-Remaining` header currently shows a value that is off by 1 (e.g., 499 instead of 500) due to the `limit-count` module deducting cost during the access-phase dry-run peek. This will be fixed in a follow-up PR after [apisix-build-tools#455](https://github.com/api7/apisix-build-tools/pull/455) merges and a new `apisix-runtime` is released with `lua-resty-limit-traffic` v1.2.0, which supports `cost=0` for non-deducting peeks.

### CODE REVIEW

File: apisix/plugins/ai-rate-limiting.lua [hunk 1]
```diff
@@ -136,8 +149,42 @@ local limit_conf_cache = core.lrucache.new({
...
 
+-- safe math functions allowed in cost expressions
+local expr_safe_env = {
+    math = math,
+    abs = math.abs,
+    ceil = math.ceil,
+    floor = math.floor,
+    max = math.max,
+    min = math.min,
+}
```
[the remark this answers is not in the document]
Comment (nic-6443) [reply apisix/plugins/ai-rate-limiting.lua:161]: Good catch. Fixed — raw usage keys that collide with expr_safe_env names are now skipped.

File: apisix/plugins/ai-rate-limiting.lua [hunk 2]
```diff
@@ -136,8 +149,42 @@ local limit_conf_cache = core.lrucache.new({
...
+
+local function compile_cost_expr(expr_str)
+    local fn_code = "return " .. expr_str
+    -- validate syntax by loading first
+    local fn, err = load(fn_code, "cost_expr", "t", expr_safe_env)
+    if not fn then
+        return nil, err
```
[the remark this answers is not in the document]
Comment (nic-6443) [reply apisix/plugins/ai-rate-limiting.lua:168]: The expression is configured by admins via the plugin config, not by end-users or API callers. Admins already have full control over the gateway. A strict arithmetic-only validator would add significant complexity. This is consistent with the upstream implementation.

File: apisix/plugins/ai-rate-limiting.lua [hunk 3]
```diff
@@ -264,7 +311,51 @@ function _M.check_instance_status(conf, ctx, instance_name)
 end
 
 
+local function eval_cost_expr(conf_cost_expr, raw)
+    local fn_code = "return " .. conf_cost_expr
```
[the remark this answers is not in the document]
Comment (nic-6443) [reply apisix/plugins/ai-rate-limiting.lua:316]: Valid optimization opportunity. The expression strings are typically very short and load() in LuaJIT is fast. This is consistent with the upstream implementation. Caching the compiled function can be done as a follow-up optimization if profiling shows it matters.

File: apisix/plugins/ai-rate-limiting.lua [hunk 4]
```diff
@@ -264,7 +311,51 @@ function _M.check_instance_status(conf, ctx, instance_name)
...
+        return nil, "failed to compile cost_expr: " .. err
+    end
+    local ok, result = pcall(fn)
+    if not ok then
+        return nil, "failed to evaluate cost_expr: " .. result
+    end
+    if type(result) ~= "number" then
+        return nil, "cost_expr must return a number, got: " .. type(result)
+    end
+    return math_floor(result + 0.5)
```
[the remark this answers is not in the document]
Comment (nic-6443) [reply apisix/plugins/ai-rate-limiting.lua:349]: Fixed — now rejects NaN/inf and clamps negative results to 0. Added test case (TEST 13) to verify.

File: apisix/plugins/ai-rate-limiting.lua [hunk 5]
```diff
@@ -61,10 +65,19 @@ local schema = {
...
         },
+        cost_expr = {
+            type = "string",
+            minLength = 1,
+            description = "Lua arithmetic expression for dynamic token cost calculation. "
+                .. "Variables are injected from the LLM API raw usage response fields. "
+                .. "Missing variables default to 0. "
+                .. "Only valid when limit_strategy is 'expression'. "
+                .. "Example: input_tokens + cache_creation_input_tokens + output_tokens",
+        },
```
[the remark this answers is not in the document]
Comment (nic-6443) [reply apisix/plugins/ai-rate-limiting.lua:81]: Having cost_expr set when strategy is not expression is harmless (ignored). Rejecting it would make config management harder (e.g., switching strategies would require removing/adding the field). This follows the be-liberal-in-what-you-accept principle.

File: t/plugin/ai-rate-limiting-expression.t [hunk 6]
```diff
@@ -0,0 +1,551 @@
...
+                    limit_strategy = "expression",
+                    cost_expr = "input_tokens + output_tokens",
+                },
+                -- valid complex expression
+                {
+                    limit = 100,
+                    time_window = 60,
+                    limit_strategy = "expression",
+                    cost_expr = "(input_tokens - cache_read_input_tokens) + cache_creation_input_tokens * 1.25 + output_tokens",
+                },
```
[the remark this answers is not in the document]
Comment (nic-6443) [reply t/plugin/ai-rate-limiting-expression.t:192]: Added TEST 12/13 — expression input_tokens - cache_read_input_tokens yields -150, which is clamped to 0. Remaining stays at 99 for both requests.

**Comment by nic-6443** [comment]
Fixed the CI failure in TEST 13: when the expression evaluates to a negative value that gets clamped to 0, calling `rate_limit()` with `cost=0` triggers an assertion failure in `resty.limit.count`'s `incoming` function (`dict:incr(key, 0, ...)`). The fix skips the `rate_limit()` call entirely when `used_tokens == 0` since there's nothing to deduct.

**Review by membphis — APPROVED** [review]

**Review by AlinsRan — APPROVED** [review]

**Review by shreemaan-abhishek — APPROVED** [review]
